tests/checks/preflight.go

```diff
@@ -38,39 +39,75 @@ type ClusterProfile struct {
 	IsKind *bool `json:"isKind,omitempty"`
 }
 
-func IntrospectCluster() {
+type clusterIntrospector struct {
+	DiscoveredClusterProfile ClusterProfile
+	ExpectedClusterProfile   ClusterProfile
+}
+
+func (c *clusterIntrospector) BeforeSuiteVerification() {
+	c.DiscoveredClusterProfile = c.introspectCluster()
+	c.ExpectedClusterProfile = c.loadExpectedClusterProfile()
+	// XXX this flags needs to go
+	// it makes dual stack the default assumption if no profile file is provided
+	if len(flags.ClusterProfilePath) == 0 {
+		deprecatedIsDualStack := !flags.SkipDualStackTests
+		c.ExpectedClusterProfile.DualNetworkStack = &deprecatedIsDualStack
+	}
+	printProfile("Discovered cluster:", &c.DiscoveredClusterProfile)
+	printProfile("Expected cluster:", &c.ExpectedClusterProfile)
+	if err := c.verifyClusterExpectations(c.ExpectedClusterProfile, c.DiscoveredClusterProfile); err != nil {
+		ginkgo.Fail(fmt.Sprintf("%v", err))
+	}
+}
+
+func (c *clusterIntrospector) introspectCluster() ClusterProfile {
+	discovered := ClusterProfile{}
 	virtClient, err := kubecli.GetKubevirtClient()
 	util.PanicOnError(err)
 	nodes := util.GetAllSchedulableNodes(virtClient).Items
 	nodeCount := len(nodes)
-	DiscoveredClusterProfile.MinimumSchedulableNodes = &nodeCount
+	discovered.MinimumSchedulableNodes = &nodeCount
 	if IsOpenShift() {
 		majorVersion := cluster.GetOpenShiftMajorVersion(virtClient)
-		DiscoveredClusterProfile.OpenShiftMajorVersion = &majorVersion
+		discovered.OpenShiftMajorVersion = &majorVersion
 	}
 	kubernetesVersion, err := cluster.GetKubernetesVersion(virtClient)
 	if err != nil {
 		util.PanicOnError(err)
 	}
-	DiscoveredClusterProfile.KubernetesVersion = &kubernetesVersion
+	discovered.KubernetesVersion = &kubernetesVersion
 
 	dualStack, err := libnet.IsClusterDualStack(virtClient)
 	if err != nil {
 		util.PanicOnError(err)
 	}
-	DiscoveredClusterProfile.DualNetworkStack = &dualStack
+	discovered.DualNetworkStack = &dualStack
 
 	withCPUManager := 0
 	for _, node := range nodes {
 		if IsCPUManagerPresent(&node) {
 			withCPUManager++
 		}
 	}
-	DiscoveredClusterProfile.MinimumNodesWithCPUManager = &withCPUManager
+	discovered.MinimumNodesWithCPUManager = &withCPUManager
 
 	isKind := IsRunningOnKindInfra() || IsRunningOnKindInfraIPv6()
-	DiscoveredClusterProfile.IsKind = &isKind
+	discovered.IsKind = &isKind
+	return discovered
+}
 
+func (c *clusterIntrospector) loadExpectedClusterProfile() ClusterProfile {
+	expected := &ClusterProfile{}
+	if len(flags.ClusterProfilePath) > 0 {
+		reader, err := os.Open(flags.ClusterProfilePath)
+		if err != nil {
+			log.DefaultLogger().Reason(err).Critical("Could not find the provided cluster profile file.")
```

Comment:
Shouldn't we be more strict here and exit the suite with error that we couldn't open the provided cluster profile?